File: appambit_sdk_flutter/android/src/main/kotlin/com/appambit/appambit_sdk_flutter/CrashesFlutter.kt
package com.appambit.appambit_sdk_flutter

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.appambit.sdk.Crashes
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodChannel

class CrashesFlutter {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context

    fun attach(binding: FlutterPlugin.FlutterPluginBinding, context: Context) {
        this.context = context
        channel = MethodChannel(binding.binaryMessenger, "com.appambit/crashes")
        channel.setMethodCallHandler { call, result ->
            when (call.method) {
                "didCrashInLastSession" -> {
                    val didCrash = Crashes.didCrashInLastSession()
                    result.success(didCrash)
                }
                "generateTestCrash" -> {
                    result.success(null)
                    Handler(Looper.getMainLooper()).post {
                        Crashes.generateTestCrash()
                    }
                }
                "logErrorMessage" -> {
                    val args = (call.arguments as? Map<*, *>) ?: emptyMap<Any, Any>()
                    val message = args["message"] as? String
                    val propertiesIn = args["properties"] as? Map<*, *>
                    val props = asStringMap(propertiesIn)
                    if (message == null) {
                        result.error("NO_MESSAGE", "Message null", null)
                    } else {
                        Crashes.logError(message, props.ifEmpty { null })
                        result.success(null)
                    }
                }
                "logError" -> {
                    val args = (call.arguments as? Map<*, *>) ?: emptyMap<Any, Any>()
                    val msg = (args["message"] as? String) ?: (args["errorMessage"] as? String) ?: "UnknownException"
                    val stack = args["stackTrace"] as? String
                    val classFqn = args["classFqn"] as? String
                    val fileName = args["fileName"] as? String
                    val lineNumber = anyToLong(args["lineNumber"])
                    val propertiesIn = args["properties"] as? Map<*, *>
                    val props = asStringMap(propertiesIn)

                    val ex = Exception(msg)
                    val frames = mutableListOf<StackTraceElement>()
                    if (classFqn != null || fileName != null || lineNumber != null) {
                        val cls = classFqn ?: "flutter"
                        val file = lastSegment(fileName ?: "unknown.dart")
                        val line = lineNumber?.toInt() ?: -1
                        frames.add(StackTraceElement(cls, "call", file, line))
                    }
                    if (!stack.isNullOrEmpty()) {
                        frames.addAll(parseDartStack(stack))
                    }
                    if (frames.isEmpty()) {
                        frames.add(StackTraceElement("flutter", "call", "unknown.dart", -1))
                    }
                    ex.stackTrace = frames.toTypedArray()

                    Crashes.logError(ex, props.ifEmpty { null })
                    result.success(null)
                }
                else -> result.notImplemented()
            }
        }
    }

    fun detach() {
        channel.setMethodCallHandler(null)
    }

    private fun asStringMap(map: Map<*, *>?): Map<String, String> {
        if (map == null) return emptyMap()
        val out = mutableMapOf<String, String>()
        for ((k, v) in map) {
            val key = k as? String ?: continue
            val value = v?.toString() ?: continue
            out[key] = value
        }
        return out
    }

    private fun anyToLong(any: Any?): Long? = when (any) {
        is Long -> any
        is Int -> any.toLong()
        is Number -> any.toLong()
        is String -> any.toLongOrNull()
        else -> null
    }

    private fun parseDartStack(stack: String): List<StackTraceElement> {
        val lines = stack.split('\n')
        val elems = mutableListOf<StackTraceElement>()
        val regex = Regex("^#\\d+\\s+([^\\s]+)\\s+\\((.+?):(\\d+)(?::(\\d+))?\\)$")
        for (raw in lines) {
            val line = raw.trim()
            if (line.isEmpty()) continue
            val m = regex.find(line) ?: continue
            val symbol = m.groupValues[1]
            val loc = m.groupValues[2]
            val ln = m.groupValues[3].toIntOrNull() ?: -1
            val className = symbol.substringBefore('.', symbol)
            val method = if (symbol.contains('.')) symbol.substringAfter('.') else symbol
            val file = lastSegment(normalizePath(loc))
            elems.add(StackTraceElement(className, method, file, ln))
        }
        return elems
    }

    private fun normalizePath(loc: String): String {
        return if (loc.startsWith("file:")) kotlin.runCatching { java.net.URI(loc).path ?: loc }.getOrDefault(loc) else loc
    }

    private fun lastSegment(path: String): String {
        val p = path.replace('\\', '/')
        val idx = p.lastIndexOf('/')
        return if (idx >= 0) p.substring(idx + 1) else p
    }
}

File: appambit_sdk_flutter/ios/Classes/CrashesFlutter.swift
import Foundation
import Flutter
import AppAmbitSdk


final class CrashesFlutter {
 
  static func didCrashInLastSession(result: @escaping FlutterResult) {
    Crashes.didCrashInLastSession { didCrash in
      result(didCrash)
    }
  }

  static func generateTestCrash(result: @escaping FlutterResult) {
    Crashes.generateTestCrash()
  }

    static func logError(args: Any?, result: @escaping FlutterResult) {
      guard let dict = args as? [String: Any]? else {
        result(FlutterError(code: "BAD_ARGS", message: "Expected Map or null", details: nil))
        return
      }

      let properties = dict?["properties"] as? [String: String]
      let classFqn   = dict?["classFqn"] as? String
      let fileName   = dict?["fileName"] as? String
      let lineNumber = (dict?["lineNumber"] as? NSNumber)?.int64Value
                       ?? Int64(dict?["lineNumber"] as? Int ?? #line)

      func errorFromExceptionMap(_ exc: [String: Any]) -> Error {
        let message = (exc["message"] as? String) ?? (exc["description"] as? String) ?? "Unknown exception"
        let code = (exc["code"] as? Int) ?? 0

        var userInfo: [String: Any] = [NSLocalizedDescriptionKey: message]

        if let stack = exc["stackTrace"] as? String {
          userInfo["stackTrace"] = stack
        }
        if let source = exc["source"] as? String {
          userInfo["source"] = source
        }
        if let inner = exc["innerException"] as? String {
          userInfo["innerException"] = inner
        }

        for (k, v) in exc {
          if userInfo[k] == nil {
            userInfo[k] = v
          }
        }

        return NSError(domain: "com.appambit.flutter.error", code: code, userInfo: userInfo)
      }

      var exceptionError: Error? = nil
      if let excMap = dict?["exception"] as? [String: Any] {
        exceptionError = errorFromExceptionMap(excMap)
      } else if let msg = dict?["message"] as? String ?? dict?["errorMessage"] as? String {
        var exc: [String: Any] = ["message": msg]
        if let stack = dict?["stackTrace"] as? String { exc["stackTrace"] = stack }
        exceptionError = errorFromExceptionMap(exc)
      } else if dict == nil {
        exceptionError = nil
      } else {
        exceptionError = nil
      }

      Crashes.logError(
        exception: exceptionError,
        properties: properties,
        classFqn: classFqn,
        fileName: fileName,
        lineNumber: lineNumber
      ) { err in
        if let err {
          result(FlutterError(code: "LOG_ERROR_ERR", message: err.localizedDescription, details: nil))
        } else {
          result(nil)
        }
      }
    }


  static func logErrorMessage(args: Any?, result: @escaping FlutterResult) {
    guard
      let dict = args as? [String: Any],
      let message = dict["message"] as? String
    else {
      result(FlutterError(code: "BAD_ARGS", message: "Missing 'message'", details: nil))
      return
    }

    let properties = dict["properties"] as? [String: String]
    let classFqn   = dict["classFqn"] as? String
    let fileName   = dict["fileName"] as? String
    let lineNumber = (dict["lineNumber"] as? NSNumber)?.int64Value
                     ?? Int64(dict["lineNumber"] as? Int ?? #line)

    Crashes.logError(
      message: message,
      properties: properties,
      classFqn: classFqn,
      fileName: fileName,
      lineNumber: lineNumber
    ) { err in
      if let err { result(FlutterError(code: "LOG_ERROR_MSG_ERR", message: err.localizedDescription, details: nil)) }
      else { result(nil) }
    }
  }
}
